### PR TITLE
Add Full-Screen size to Modal and Responsive variations for breakpoints

### PR DESCRIPTION
**Pull request by GeoSot**
Add Full-Screen size to Modal and Responsive variations for breakpoints
Works from breakpoint and down

( @MartijnCuppens , please ignore my previous pull request)

Preview: <https://deploy-preview-28683--twbs-bootstrap.netlify.com/docs/4.3/components/modal/#fullscreen-modal>
Previous PR: #28682

### CODE REVIEW

**Comment by MartijnCuppens** [comment]
Yu @GeoSot,

Nice to see you getting involved in Bootstrap! I've readded the netlify preview in your (you can get the link in the PR tests). Try not to reopen too many PR's for the same feature, we can miss some essential information sometimes. 

I noticed there's a spacing at the right side of the screen, could you fix that?

**Comment by GeoSot** [comment]
@MartijnCuppens you are right about the two PR's. (I 've messed up the commits and wanted a clear solution)

As for the padding on the right, I saw that it is done from js. 
As a solution I can change max-width to min-width, in order to have a nice optical result, but is it acceptable?

**Comment by MartijnCuppens** [comment]
Hmmm, I can't think of another better solution right now. Maybe @ysds knows a better way of handling this?

**Comment by ysds** [comment]
I’ll take a look later :)

**Comment by ysds** [comment]
As far as looked through that, I felt three things:

- The class name is a bit confusing. e.g. I misunderstood the `.modal-md-full` makes the modal to be full screen when the screen is wide.

- There is a display problem when the full screen modal has a lot of content.

- The modal's right padding is for centering the modal to the page content not the window (#14892). But the full screen modal will not require the padding. To remove the padding, should change the JS because it is added by JS.

**Comment by GeoSot** [comment]
@MartijnCuppens  I have made some changes and fixes according to the previous suggestions. 

As for the className, i am open to  suggestions

File: site/content/docs/4.3/components/modal.md
Comment (XhmikosR) [inline site/content/docs/4.3/components/modal.md]: Please make sure you use 2-space indentation in all of your changes.

**Comment by MartijnCuppens** [comment]
> As for the className, i am open to suggestions

`fullscreen` would make more sense I guess.

There's something else. We use a [mobile-first](https://getbootstrap.com/docs/4.3/layout/overview/#responsive-breakpoints) approach in Bootstrap, which uses `min-width` media queries instead of `max-width`. We also need extra documentation and examples to show the responsive effect.

**Comment by GeoSot** [comment]
> There's something else. We use a [mobile-first](https://getbootstrap.com/docs/4.3/layout/overview/#responsive-breakpoints) approach in Bootstrap, which uses `min-width` media queries instead of `max-width`. We also need extra documentation and examples to show the responsive effect.


I have changed the class name. 

As for the mobile first approach thing, the purpose is to have a Full Screen modal on mobile , till the chosen breakpoint. So I am not sure if it needs any change

**Comment by XhmikosR** [comment]
@GeoSot something seems broken. Also please squash your patches into one after rebasing against upstream.

**Comment by GeoSot** [comment]
@XhmikosR  can I do something more in order to help this pull-request to proceed?

**Comment by mdo** [comment]
Trying this out myself, I don't think the fullscreen modal should "drop in" like the other modals—gives it a weird looking animation. Thoughts on revisiting that part?

**Comment by GeoSot** [comment]
Maybe if we omit the 'fade' class. Or try another transition???

**Comment by MartijnCuppens** [comment]
I would keep the fade effect. Maybe we split the fade effect and the slide-in effect into 2 classes? Better cover this in another PR then because all modal classes.

**Comment by GeoSot** [comment]
So, what you propose me to do?

**Comment by MartijnCuppens** [comment]
> So, what you propose me to do?

Interesting question. I know @ysds is working on splitting the transition classes from the components in https://github.com/twbs/bootstrap/pull/29081.

@mdo, maybe we should postpone the `drop in` change until https://github.com/twbs/bootstrap/pull/29081 is finished?

**Comment by mdo** [comment]
Sounds fine by me!

**Comment by MartijnCuppens** [comment]
I'm still a bit concerned this implementation will be too confusing, other responsive components work the other way around:

- `.col-md-4`: 4 columns **above** `md` breakpoint
- `.navbar-expand-md`: expand **above** `md` breakpoint
- ...

This implementation:
- `.modal-md-fullscreen`: fullscreen modal **below** `md` breakpoint

Maybe we should have a look at responsive modal sizing in general first?

**Comment by GeoSot** [comment]
> I'm still a bit concerned this implementation will be too confusing, other responsive components work the other way around:

I understand your concern and the logic of it.

The only thing I can say as simple user is, that a fullscreen modal seems more reasonable on small screen sizes than on wide screens. As for the naming conversion, if anyone want to suggest another approach I am open to think of it

File: scss/_modal.scss
Comment (MartijnCuppens) [inline scss/_modal.scss]: I think we can safely move this to the global `.modal-body`, would generate less code.

File: scss/_modal.scss
Comment (MartijnCuppens) [inline scss/_modal.scss]: I think we can safely move the height to the global `.modal-content`, would generate less code.

File: scss/_modal.scss
Comment (MartijnCuppens) [inline scss/_modal.scss]: Why `inherit`?

Comment (GeoSot) [reply scss/_modal.scss]: Because .modal-dialog has a max-width:500px and has to be overwritten

Comment (MartijnCuppens) [reply scss/_modal.scss]: Use `none` then, that's the default value. `inherit` implies it **needs** to be inherited.

Comment (GeoSot) [reply scss/_modal.scss]: @MartijnCuppens  Done

**Comment by Glenn-Domin** [comment]
I'd definitely be confused by `.modal-md-fullscreen` applying to `md` and down where everything else in Bootstrap is `md` and up. Not sure how practical this is but if you wanted to do a mobile-only implementation I'd suggest doing something like `.modal-fullscreen .modal-lg-boxed` to make it not full from `lg` upward (I don't like the name "boxed" but it's off the top of my head to demonstrate the point).

**Comment by ysds** [comment]
The implementation with media-break-down would be an appropriate way here. But the issue is the class naming. I searched for past considerations about class naming of the similar case, found the [hidden-xx-down](https://v4-alpha.getbootstrap.com/layout/responsive-utilities/#available-classes) and the [btn-block-xx-down](https://github.com/twbs/bootstrap/pull/21720#issuecomment-285197633). I think such naming makes sense here too. So my suggestion is `.modal-fullscreen` (all viewports) and `.modal-fullscreen-md-down` ...

Regarding to the animation, implementation of #29081 still takes time. so how about leaving it as is for now for proceed this PR?

**Comment by MartijnCuppens** [comment]
Brilliant plan @ysds. @GeoSot, could change the implementation to @ysds' suggestion?

File: site/content/docs/4.3/components/modal.md
Comment (nteetor) [inline site/content/docs/4.3/components/modal.md]: Chanage to `.modal-fullscreen-xl-down`.

File: site/content/docs/4.3/components/modal.md
Comment (nteetor) [inline site/content/docs/4.3/components/modal.md]: Change to `.modal-fullscreen-lg-down`.

File: site/content/docs/4.3/components/modal.md
Comment (nteetor) [inline site/content/docs/4.3/components/modal.md]: `.modal-fullscreen-md-down`

File: site/content/docs/4.3/components/modal.md
Comment (nteetor) [inline site/content/docs/4.3/components/modal.md]: `.modal-fullscreen-sm-down`

**Comment by nteetor** [comment]
I noticed the names were off in the list of classes under the Fullscreen section. I dropped in a few comments pointing out which classes needed to be corrected.

**Comment by ysds** [comment]
Since there are still some issues, I will add the commit directly.

Adding `overflow-y: auto` to the global `.modal-body` will increase encountering #28513 (See [demo](https://codepen.io/fellows3/pen/xxGdwra)). Maybe should handle with JavaScript, but for workaround here, moving it to the each `.modal-body`.

**Comment by andreivictor** [comment]
Just my humble opinion:

- looking at the generated css, we have:
`@media (max-width: 767.98px)
.modal-fullscreen-sm-down { ... }`

- but in the documentation it says:
`.modal-fullscreen-sm-down | Below 576px`

By using the same logic from https://v4-alpha.getbootstrap.com/layout/responsive-utilities/#available-classes, `modal-fullscreen-*-down` would mean that the modal is fullscreen at the given breakpoint or smaller.

**Comment by ysds** [comment]
I added changes:
* Fix header and footer shrink issue
* Move `overflow-y: auto` from the global `.modal-body` to each fullscreens
* Fix wrong breakpoint postfixs and remove modalFullscreen() mixin
* Improve documentation

**Review by ysds — APPROVED** [review]
LGTM, but need another approval from @twbs/css-review

**Review by MartijnCuppens — APPROVED** [review]
I've added border radii resets to the footer and header, just in case a background color is added. Thanks all for contributing to feature! ❤️

**Comment by MartijnCuppens** [comment]
@XhmikosR, how are we gonna merge this? Squash it into one commit?

**Comment by XhmikosR** [comment]
This should probably me manually rebased and squash the patches per author.

**Comment by XhmikosR** [comment]
Unless you guys don't mind squashing everything into one when merging (with a clean commit message, though)

**Comment by MartijnCuppens** [comment]
I personally don't care. We could maybe add Co-authors in the commit message:

```
Co-authored-by: XhmikosR <xhmikosr@gmail.com>
Co-authored-by: Shohei Yoshida <fellows3@gmail.com>
Co-authored-by: Martijn Cuppens <martijn.cuppens@gmail.com>
```

**Comment by XhmikosR** [comment]
@ysds can you have another look after the rebase?

@GeoSot please make sure you don't overwrite my rebase 🙂

**Comment by yadoudou** [comment]
unsubscribed



---Original---

**Review by ysds — APPROVED** [review]
All looks good to me 👍

**Pull request merged**
